File: examples/governance_static_go/domain/monitoring.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type Monitoring struct {
	ID        string    `json:"id"`
	ModelId string `json:"model_id"`
	DeploymentId string `json:"deployment_id"`
	MetricName string `json:"metric_name"`
	Value float64 `json:"value"`
	Threshold float64 `json:"threshold"`
	RecordedAt time.Time `json:"recorded_at"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func NewMonitoring(modelId string, deploymentId string, metricName string, value float64, threshold float64, recordedAt time.Time) *Monitoring {
	a := &Monitoring{
		ID:        uuid.New().String(),
		ModelId: modelId,
		DeploymentId: deploymentId,
		MetricName: metricName,
		Value: value,
		Threshold: threshold,
		RecordedAt: recordedAt,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *Monitoring) Validate() error {
	if a.ModelId == "" {
		return &ValidationError{Field: "model_id", Message: "model_id can't be blank"}
	}
	if a.MetricName == "" {
		return &ValidationError{Field: "metric_name", Message: "metric_name can't be blank"}
	}
	// invariant: threshold must be positive
	return nil
}

File: examples/governance/operations_domain/lib/operations_domain/monitoring/monitoring.rb
require 'hecks/mixins/model'

module OperationsDomain
  class Monitoring
    include Hecks::Model

    attribute :model_id
    attribute :deployment_id
    attribute :metric_name
    attribute :value
    attribute :threshold
    attribute :recorded_at

    private

    def validate!
      raise ValidationError.new("model_id can't be blank", field: :model_id, rule: :presence) if model_id.nil? || (model_id.respond_to?(:empty?) && model_id.empty?)
      raise ValidationError.new("metric_name can't be blank", field: :metric_name, rule: :presence) if metric_name.nil? || (metric_name.respond_to?(:empty?) && metric_name.empty?)
    end

    def check_invariants!
      raise InvariantError, "threshold must be positive" unless instance_eval(&proc { !threshold || threshold > 0 })
    end
  end
end
